File: src/FASTCAMPUS/ALGO/D15/P15649.js
const fs = require('fs');
class Scanner {
  read
  stringToken
  constructor(fileName){
    this.read = fs.readFileSync(fileName).toString().split('\n');
    this.stringToken = [];
  }
  next(){
    if(this.stringToken.length === 0){
      if(this.read.length > 0){
        this.stringToken = this.read.shift().split(" ");
      } else {
        return null;
      } 
    }
    return this.stringToken.shift(); 
  }

  nextNumber(){
    const next = this.next();
    if(next){
      return Number(next);
    } else {
      return null;
    }
    
  }

  nextLine(){
    if(this.read.length > 0){
      return this.read.shift();
    } else {
      return null;
    }
    
  }
  
}

let N, M;
const selected = [0,];
let stringBuilder = "";
//1에서 N 까지 자연수 중 M개를 고른 수열 
//중복불가 / 순서가 다르면 다른 케이스 
//시간복잡도 o(N!/(N-M)!)
const input = function(){
  const scanner = new Scanner('./stdin.txt');
  N = scanner.nextNumber();
  M = scanner.nextNumber();
}

const rec_func = function(k){
  if( k === M + 1){
    for(let i = 1; i < k; i ++){
      stringBuilder += (selected[i] + " ");
    }
    stringBuilder += `\n`;
  } else {
    for( let i = 1; i <= N; i ++ ){
      if(k > 1){
        if(selected.indexOf(i) !== -1){
          continue;
        }
      } 
      selected.push(i);
      rec_func(k + 1);
      selected.pop();
    }
  }
}
input();
rec_func(1);
console.log(stringBuilder);

File: src/FASTCAMPUS/ALGO/D15/P15649.java
package FASTCAMPUS.ALGO.D15;
import java.io.*;
import java.util.*;
//o(N!/(N-M)!)
public class P15649 {
    static int N, M;
    static ArrayList<Integer> selected;
    static StringBuilder sb;
    static void input() {
        FastReader scan = new FastReader();
        N = scan.nextInt();
        M = scan.nextInt();
        selected = new ArrayList<Integer>(M);
        sb = new StringBuilder();
    }
    public static void main(String[] args) {
        input();
        rec_func(1);
        
    }
    public static void rec_func(int k){
      if( k == M + 1){
        System.out.println(selected);
        return;
      } else {
        for(int cand = 1; cand <= N; cand ++){
          if(selected.indexOf(cand) == -1){
            selected.add(cand);
            rec_func(k + 1);
            selected.remove(selected.size() - 1);
          }
            
        }
      }
    }
    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        public FastReader(String s) throws FileNotFoundException {
            br = new BufferedReader(new FileReader(new File(s)));
        }

        String next() {
            while (st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }

        long nextLong() {
            return Long.parseLong(next());
        }

        double nextDouble() {
            return Double.parseDouble(next());
        }

        String nextLine() {
            String str = "";
            try {
                str = br.readLine();
            } catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }
}